---
title: "Bagel_Scratch"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
bay = readRDS(system.file("testdata", "TODO_BAGEL.rds", package = "BAGEL"))

vcf_file <- system.file("testdata", "public_LUAD_TCGA-97-7938.vcf", 
                         package = "BAGEL")
#vcf <- VariantAnnotation::readVcf(vcf_file)
#vcf <- VariantAnnotation::readVcfAsVRanges(vcf_file)
vcf_dt <- vcf_file_to_dt(vcf_file)



vcf <- VariantAnnotation::VRanges(seqnames = vcf_dt$Chromosome, ranges = IRanges(vcf_dt$Start_Position, vcf_dt$End_Position), ref = vcf_dt$Tumor_Seq_Allele1, alt = vcf_dt$Tumor_Seq_Allele2)

#Reference bases are always stranded exactly one way, so for 
#determining transcription strand, we don't have to worry about reverse complements
#refs <- BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19
#bs_overlaps <- BSgenome::getSeq(refs, vcf)

library("TxDb.Hsapiens.UCSC.hg19.knownGene")
genes_hg19 <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
overlaps <- findOverlaps(vcf, genes_hg19)
head(overlaps)
transcribed_variants = rep("NA", length(vcf))

transcribed_variants[queryHits(overlaps)] = as.character(decode(strand(genes_hg19[subjectHits(overlaps)])))

vcf_dt$Transcribed_Strand <- transcribed_variants
a=new("bagel", variants = vcf_dt)
BAGEL:::build_custom_table(a, "Transcribed_Strand", "transcribed_strand")
a@variants[-which(a@variants$Transcribed_Strand == "NA"), ]
drop_na_variants <- function(bay, annotation) {
  bay@variants[-which(bay@variants[[annotation]] == "NA"), ]
  
}

drop_na_variants(bay@variants, "Transcribed_Strand")

subset_
res = discover_signatures(a, "transcribed_strand", 3, "nmf")
plot_signatures(res)


#variants = cbind(as.character(decode(seqnames(vcf))), start(ranges(vcf)), VariantAnnotation::ref(vcf), VariantAnnotation::alt(vcf), as.character(decode(VariantAnnotation::sampleNames(vcf))))
variants_stranded = cbind(variants, transcribed_variants)
only_matched = variants_stranded[variants_stranded[, "transcribed_variants"] %in% c("+", "-"), ]
table(only_matched[, 6])
table(paste0(only_matched[, 3], ">", only_matched[, 4], "_", only_matched[, 6]))
```

```{r}
annotate_transcript_strand <- function(bay, genome_build, build_table = TRUE) {
  if (genome_build %in% c("19", "hg19")) {
    genes <- genes(
      TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene)
  } else if (genome_build %in% c("38", "hg38")) {
    genes <- genes(
      TxDb.Hsapiens.UCSC.hg38.knownGene::TxDb.Hsapiens.UCSC.hg38.knownGene)
  } else if (methods::isClass(genome_build, "TxDb")) {
    genes <- genome_build
  } else {
    stop("Please select hg19, hg38, or provde a TxDb object")
  }
  
  dat <- bay@variants
  
  #Create VRanges object to determine strand of variants within genes
  vrange <- VariantAnnotation::VRanges(seqnames = dat$Chromosome, ranges = 
                                      IRanges(dat$Start_Position, 
                                              dat$End_Position), ref = 
                                      dat$Tumor_Seq_Allele1, alt = 
                                      dat$Tumor_Seq_Allele2)
  overlaps <- IRanges::findOverlaps(vrange, genes)
  transcribed_variants = rep("NA", length(vrange))
  transcribed_variants[S4Vectors::queryHits(overlaps)] = 
    as.character(S4Vectors::decode(GenomicRanges::strand(
      genes[S4Vectors::subjectHits(overlaps)])))
  
  dat[[Transcribed_Strand]] <- transcribed_variants
  eval.parent(substitute(bay@variants <- dat))
  if (build_table) {
    dat_bagel <- methods::new("bagel", variants = dat, count_tables =
                               bay@count_tables,
                             sample_annotations = bay@sample_annotations)
    tab <- build_custom_table(dat_bagel, variant_annotation = Transcribed_Strand,
                         name = Transcribed_Strand, return_instead = FALSE)
    eval.parent(substitute(bay@count_tables <- tab))
  }
}
```

```{r}
#Make sure transcript strand was done correctly, compare to MutationalPatterns stranding
vcfs <- readRDS(system.file("states/read_vcfs_as_granges_output.rds",
                package="MutationalPatterns"))
ref_genome = "BSgenome.Hsapiens.UCSC.hg19"
library(ref_genome, character.only = TRUE)
genes_hg19 <- readRDS(system.file("states/genes_hg19.rds",
                        package="MutationalPatterns"))

mut_mat_s = MutationalPatterns::mut_matrix_stranded(vcfs, ref_genome, genes_hg19, 
                                mode = "transcription")


mv = vcfs@unlistData
vcf_names <- names(vcfs)
mv_dats <- list()
for(name in vcf_names) {
  mv = vcfs[[name]]
  mv_dats[[name]] <- data.table::data.table(Chromosome = as.character(seqnames(mv)), Start_Position = start(ranges(mv)), End_Position = end(ranges(mv)), Tumor_Seq_Allele1 = as.character(mv@elementMetadata@listData$REF), Tumor_Seq_Allele2 =
                                   sapply(mv@elementMetadata@listData$ALT, function(x) substr(as.character(x[[1]]), 1, 1)), Tumor_Sample_Barcode = name)
}
mutpatdat <- do.call(rbind, mv_dats)
mut_bay <- new('bagel', variants = mutpatdat)
annotate_variant_type(mut_bay)
annotate_transcript_strand(mut_bay, 19)
g = select_genome("19")
create_snv192_table(mut_bay, g)

dat

substr(as.character(mv@elementMetadata@listData$ALT[[1]]), 1, 1)


```

```{r}
graph <- read.table("/Users/atgc/Downloads/GSM4305465_WT_siLIG1_rep1_rfd_10kb.bedgraph", stringsAsFactors = FALSE)
colnames(graph) <- c("chr", "start", "end", "RFD")

#This order is important because the word "leading" > 0 because R
graph[which(graph$RFD > 0), "RFD"] = "lagging"
graph[which(graph$RFD < 0), "RFD"] = "leading"
graph = graph[-which(graph$RFD == 0), ]

dat <- bay@variants
#Create VRanges object to determine strand of variants within genes
vrange <- VariantAnnotation::VRanges(seqnames = dat$Chromosome, ranges = 
                                    IRanges(dat$Start_Position, 
                                            dat$End_Position), ref = 
                                    dat$Tumor_Seq_Allele1, alt = 
                                    dat$Tumor_Seq_Allele2)
overlaps <- IRanges::findOverlaps(vrange, graph_ranges)
repl_variants = rep("NA", length(vrange))
repl_variants[S4Vectors::queryHits(overlaps)] = 
    graph_ranges$RFD[S4Vectors::subjectHits(overlaps)]

dat[["Replication_Timing"]] <- repl_variants
snv = subset_variant_by_type(drop_na_variants(dat, "Replication_Timing"), "SNV")

```

#UMAP + TSNE Origimed + PanLung
```{r}
bagel = readRDS("/Users/atgc/Documents/Grants/ZerG/annotated_5plus_counts_bagel_updated.rds")
mel = subset_bagel_by_annotation(bagel, "Tumor_Type", "Melanoma")
#TODO Combine bagels function
sclc = subset_bagel_by_annotation(bagel, "Tumor_Type", "Small_Cell_Lung_Cancer")
nsclc = subset_bagel_by_annotation(bagel, "Tumor_Type", "Non_Small_Cell_Lung_Cancer")
comb = new("bagel", variants = rbind(mel@variants, sclc@variants, nsclc@variants), sample_annotations = rbind(mel@sample_annotations, sclc@sample_annotations, nsclc@sample_annotations))
g = select_genome("19")
build_standard_table(comb, g, "SNV96")
BAGEL::cosmic_v2_subtype_map("lung")
BAGEL::cosmic_v2_subtype_map("melan")
res = predict_exposure(comb, "SNV96", cosmic_v2_sigs, signatures_to_use = c(1,2,4,5,6,7,11,13,15,17))
umap = .calculateUmap(t(res@samples), 2, minDist = 1.75, spread = 2, pca = TRUE);plot(umap)
tsne = .calculateTsne(t(res@samples), doPca = TRUE); plot(tsne)
plot_tsne = function(result, seed = 12345, perplexity = 20, initialDims = 50) {
  withr::with_seed(seed, tsne <- .calculateTsne(t(result@samples), doPca = FALSE, perplexity, initialDims = initialDims))
  tsne_dat = data.frame(x = tsne[,1], y = tsne[,2], type = result@bagel@sample_annotations$Tumor_Type)
  print(ggplot(tsne_dat, aes(x=x, y=y, col=type)) + geom_point() + ggplot2::ggtitle("Tsne") + theme_bw())
  return(tsne_dat)
}

plot_umap = function(result, seed = 12345, nNeighbors = 30, minDist = 0.75, spread = 1, pca = FALSE) {
  withr::with_seed(seed, umap <- .calculateUmap(t(result@samples), nNeighbors, minDist = minDist, spread = spread, pca = pca))
  umap_dat = data.frame(x = umap[,1], y = umap[,2], type = result@bagel@sample_annotations$Tumor_Type)
  print(ggplot(umap_dat, aes(x=x, y=y, col=type)) + geom_point() + ggplot2::ggtitle("Umap") + theme_bw())
  return(umap_dat)
}

maf_luad = auto_to_bagel_dt("/Users/atgc/Downloads/gdc_download_20200524_215820.269924/0458c57f-316c-4a7c-9294-ccd11c97c2f9/TCGA.LUAD.mutect.0458c57f-316c-4a7c-9294-ccd11c97c2f9.DR-10.0.somatic.maf", filter = FALSE, only_snp = TRUE)
maf_lusc = auto_to_bagel_dt("/Users/atgc/Downloads/gdc_download_20200524_215820.269924/95258183-63ea-4c97-ae29-1bae9ed06334/TCGA.LUSC.mutect.95258183-63ea-4c97-ae29-1bae9ed06334.DR-10.0.somatic.maf", filter = FALSE, only_snp = TRUE)
maf_mel = auto_to_bagel_dt(("/Users/atgc/Downloads/gdc_download_20200524_214639.669044/4b7a5729-b83e-4837-9b61-a6002dce1c0a/TCGA.SKCM.mutect.4b7a5729-b83e-4837-9b61-a6002dce1c0a.DR-10.0.somatic.maf"))
bagel = new("bagel", variants = maf_luad)
g=select_genome("38")
build_standard_table(bagel, g, "SNV96")
bagel_mel = new("bagel", variants = rbind(maf_luad, maf_lusc, maf_mel))
g=select_genome("38")
build_standard_table(bagel_mel, g, "SNV96")
init_sample_annotations(bagel_mel)
annot = data.frame(Samples = bagel_mel@sample_annotations, Tumor_Type = c(rep("LUAD", length(unique(maf_luad$Tumor_Sample_Barcode))), rep("LUSC", length(unique(maf_lusc$Tumor_Sample_Barcode))), rep("MEL", length(unique(maf_mel$Tumor_Sample_Barcode)))))
add_sample_annotations(bagel_mel, annot, "Samples")
panlung_mel = predict_exposure(bagel_mel, "SNV96", cosmic_v2_sigs, signatures_to_use = c(1,2,4,5,6,7,11,13,15,17))
tsne = plot_tsne(panlung_mel)
umap = plot_umap(panlung_mel)
sig7 = panlung_mel@samples[6,]
sig4 = panlung_mel@samples[3,]
sig7_log = log(sig7)
sig4_log = log(sig4)
umap$sig7_log = sig7_log
umap$sig4_log = sig4_log
umap$sig4 = sig4
ggplot(umap, aes(x=x, y=y, col=sig7, shape = type)) + geom_point() + ggplot2::ggtitle("Umap") + theme_bw()


tsne$sig7_log = sig7_log
tsne$sig7 = sig7
tsne$sig4_log = sig4_log
tsne$sig4 = sig4
ggplot(tsne, aes(x=x, y=y, col=sig4_log, shape = type)) + geom_point() + ggplot2::ggtitle("Tsne") + theme_bw()


props = panlung_mel@samples
props = sweep(props, 2, colSums(props), FUN = "/")
umap$prop2 = props[2,]
umap$prop4 = props[3,]
umap$prop5 = props[4,]
umap$prop6 = props[5,]
umap$prop7 = props[6,]
umap$prop13 = props[8,]
umap$prop15 = props[9,]

tsne$prop2 = props[2,]
tsne$prop4 = props[3,]
tsne$prop5 = props[4,]
tsne$prop6 = props[5,]
tsne$prop7 = props[6,]
tsne$prop13 = props[8,]
tsne$prop15 = props[9,]

len = nrow(umap)
umap_df = data.frame(x = rep(umap$x, 7), y = rep(umap$y, 7), type = rep(umap$type, 7), sig = c(rep(2, len), rep(4, len), rep(5, len), rep(6, len), rep(7, len), rep(13, len), rep(15, len)), sig_prop = c(umap$prop2, umap$prop4, umap$prop5, umap$prop6, umap$prop7, umap$prop13, umap$prop15))
ggplot(umap_df, aes(x=x, y=y, col=sig_prop, shape = type)) + geom_point() + facet_grid(~sig) + ggplot2::ggtitle("Umap") + theme_bw()

tsne_df = data.frame(x = rep(tsne$x, 7), y = rep(tsne$y, 7), type = rep(tsne$type, 7), sig = c(rep(2, len), rep(4, len), rep(5, len), rep(6, len), rep(7, len), rep(13, len), rep(15, len)), sig_prop = c(tsne$prop2, tsne$prop4, tsne$prop5, tsne$prop6, tsne$prop7, tsne$prop13, tsne$prop15))
ggplot(tsne_df, aes(x=x, y=y, col=sig_prop, shape = type)) + geom_point() + facet_grid(~sig) + ggplot2::ggtitle("Tsne") + theme_bw()
```

#Proportional input instead of counts for umap tsne
```{r}
prop_panlung_mel = panlung_mel
samples = panlung_mel@samples
prop_panlung_mel@samples = sweep(samples, 2, colSums(samples), FUN = "/")

prop_tsne = plot_tsne(prop_panlung_mel)
prop_umap = plot_umap(prop_panlung_mel)
prop_umap = plot_umap(prop_panlung_mel, seed= 7);ggplot(prop_umap, aes(x=x, y=y, col=type)) + geom_point() + ggplot2::ggtitle("Umap") + theme_bw()
```

```{r}
maf_luad = auto_to_bagel_dt("/Users/atgc/Downloads/gdc_download_20200524_215820.269924/0458c57f-316c-4a7c-9294-ccd11c97c2f9/TCGA.LUAD.mutect.0458c57f-316c-4a7c-9294-ccd11c97c2f9.DR-10.0.somatic.maf", filter = FALSE, only_snp = FALSE)
maf_lusc = auto_to_bagel_dt("/Users/atgc/Downloads/gdc_download_20200524_215820.269924/95258183-63ea-4c97-ae29-1bae9ed06334/TCGA.LUSC.mutect.95258183-63ea-4c97-ae29-1bae9ed06334.DR-10.0.somatic.maf", filter = FALSE, only_snp = FALSE)
maf_mel = auto_to_bagel_dt(("/Users/atgc/Downloads/gdc_download_20200524_214639.669044/4b7a5729-b83e-4837-9b61-a6002dce1c0a/TCGA.SKCM.mutect.4b7a5729-b83e-4837-9b61-a6002dce1c0a.DR-10.0.somatic.maf"), filter = FALSE, only_snp = FALSE)
bagel_mel = new("bagel", variants = rbind(maf_luad, maf_lusc, maf_mel))
g=select_genome("38")
build_standard_table(bagel_mel, g, "DBS")
```

```{r}
umap$prop1 = props[1,]
umap$prop2 = props[2,]
umap$prop4 = props[3,]
umap$prop5 = props[4,]
umap$prop6 = props[5,]
umap$prop7 = props[6,]
umap$prop11 = props[7,]
umap$prop13 = props[8,]
umap$prop15 = props[9,]
umap$prop17 = props[10,]

len = nrow(umap)
umap_df = data.frame(x = rep(umap$x, 10), y = rep(umap$y, 10), type = rep(umap$type, 10), sig = c(rep(1, len), rep(2, len), rep(4, len), rep(5, len), rep(6, len), rep(7, len), rep(11, len), rep(13, len), rep(15, len), rep(17, len)), sig_prop = c(umap$prop1, umap$prop2, umap$prop4, umap$prop5, umap$prop6, umap$prop7, umap$prop11, umap$prop13, umap$prop15, umap$prop17))

umap_df = data.frame(x = rep(umap$x, 1), y = rep(umap$y, 1), type = rep(umap$type, 1), sig = c(rep(7, len)), sig_prop = c(umap$prop7))
ggplot(umap_df, aes(x=x, y=y, col=sig_prop, shape = type)) + geom_point() + facet_wrap(~sig) + ggplot2::ggtitle("Umap") + theme_bw()
```

