---
title: "BAGEL_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{BAGEL_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library("BAGEL")
```

# Introduction

# Installation

# Setting up a bagel object
Add some text here describing what the user needs to do.

## Extracting varaints
```{r}

# Extract variants from a MAF File
lusc_maf <- system.file("testdata", "public_TCGA.LUSC.maf", package = "BAGEL") 
lusc.variants <- extract_variants_from_maf_file(maf_file = lusc_maf)

# Extract variants from an individual VCF file
luad_vcf <- system.file("testdata", "public_LUAD_TCGA-97-7938.vcf", 
                         package = "BAGEL")
luad.variants <- extract_variants_from_vcf_file(vcf_file = luad_vcf)


# Extract variants from multiple files and/or objects
melanoma_vcfs <- list.files(system.file("testdata", package = "BAGEL"), 
                           pattern = glob2rx("*SKCM*vcf"), full.names = TRUE)
variants <- extract_variants(c(lusc_maf, luad_vcf, melanoma_vcfs), verbose = TRUE)

```

## Choosing genome
Briefly describe BSgenome package and objects
```{r}
g = select_genome("hg38")
```

## Create bagel object
```{r}
bagel = create_bagel(x = variants, genome = g)
```

# Create mutation count tables
Description of what motifs are and why they need to be counted.

```{r}
build_standard_table(bagel, "SNV96")

# Generate signatures/weights
# We can use the bagel object we generated tables for using the SNV96 table
result = discover_signatures(input = bagel, table_name = "SNV96", 
                              num_signatures = 4, method = "lda", nstart = 10)


```

# Plotting
## Signatures
```{r}
##Plot results
plot_signatures(result)

# We can name signatures based on prior knowledge
name_signatures(result, c("unknown", "smoking", "apobec", "uv"))

```

## Exposures
```{r}
plot_exposures(result, proportional = TRUE)
plot_exposures(result, proportional = FALSE)
plot_sample_counts(bagel, "SNV96", get_sample_names(bagel)[1])
```

## Comparison to external signatures (e.g. COSMIC)

```{r}
##Compare to COSMIC signatures by leaving the second result as default
compare_cosmic_v2(result, threshold = 0.8)

#List which signatures correspond to subtypes including "lung"
cosmic_v2_subtype_map("lung")

#Calculate posterior based on COSMIC signatures 4, 11, 12, 15
cosmic_post = predict_exposure(bagel = bagel, "SNV96", signature_res = 
                                 cosmic_v2_sigs, signatures_to_use = 
                                 c(12, 4, 11, 15), algorithm = "lda")
#cosmic_post = predict_exposure(bagel = bagel)
#to_use = as.numeric(substr(names(which(rowSums(cosmic_post@exposures) > 10^-10)), 11, 12))
#to_use = as.numeric(substr(names(which(apply(cosmic_post@exposures, 1, max) > 10^-10)), 11, 12))
#cosmic_post = predict_exposure(bagel = bagel, signatures_to_use = to_use)

#Calculate posterior based on all calculated signatures

```

# Predicting exposures using existing signatures
```{r}
our_sigs_post = predict_exposure(bagel = bagel, "SNV96", signature_res = 
                                   result, algorithm = "lda")

#Plot results from posterior calculation
plot_signatures(cosmic_post)
plot_exposures(cosmic_post, proportional = TRUE)

plot_signatures(our_sigs_post)
plot_exposures(our_sigs_post, proportional = TRUE)

#Compare posterior results to each other
compare_results(cosmic_post, our_sigs_post, threshold = 0.70)


############################################## Error up here ^^^^^
#sample_annotations <- read.table(system.file("testdata", "sample_annotations.txt", package = "BAGEL"), sep = "\t", header=TRUE)
#init_sample_annotations(bagel)
#add_sample_annotations(bagel, sample_annotations, sample_column = "Sample_Names", columns_to_add = "Tumor_Subtypes")
#
##Add tumor type annotations to our samples
#res <- discover_signatures(bagel, table_name = "SNV96", num_signatures = 3)
#res <- readRDS(system.file("testdata", "test_res.rds", package = "BAGEL"))
#plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes")
#plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", proportional = FALSE)
#plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", by_group = FALSE)
#plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", by_group = FALSE, proportional = FALSE)

#Discover stranded SNV signatures
#brca = maf_file_to_dt("/Users/atgc/Downloads/gdc_download_20200609_120830.134740/995c0111-d90#b-4140-bee7-3845436c3b42/TCGA.BRCA.mutect.995c0111-d90b-4140-bee7-3845436c3b42.DR-10.0.somati#c.maf")
#bagel = new("bagel", variants = brca)
#
#BAGEL::build_standard_table(bagel, g, "SNV96")
#BAGEL::annotate_variant_type(bagel)
#BAGEL::annotate_transcript_strand(bagel, 38)
#BAGEL::annotate_replication_strand(bagel, BAGEL::rep_range)
#BAGEL::build_standard_table(bagel, g, "SNV192", "Transcript_Strand")
#BAGEL::build_standard_table(bagel, g, "SNV192", "Replication_Strand")
#
#####TEST WITH PANLUNG/MELANOMA
#standard <- BAGEL::discover_signatures(bagel, "SNV96", 4, "nmf")
#compare_cosmic_v2(standard, threshold = 0.9, result_name = "BRCA")
#
#trans_192 <- BAGEL::discover_signatures(bagel, "SNV192_Trans", 4, "nmf")
#plot_signatures(trans_192, plotly = TRUE)
#
#rep_192 <- BAGEL::discover_signatures(bagel, "SNV192_Rep", 4, "nmf")
#plot_signatures(rep_192, plotly = TRUE)
```

# Other functions
```{r}
#Display what samples were added
samples = get_sample_names(bagel)
```
