---
title: "BAGEL_vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{BAGEL_vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r}

#Find Signatures examples
#' examples
#' bay <- readRDS(system.file("testdata", "bagel.rds", package = "BAGEL"))
#' g <- select_genome("19")
#' create_snv96_table(bay, g)
#' res <- find_signatures(bay, "SNV96", num_signatures = 4)


#sample_annotations <- cbind(c("TCGA-94-7557-01A-11D-2122-08", "TCGA-56-7582-01A-11D-2042-08", #"TCGA-77-7335-01A-11D-2042-08", "public_SKCM_TCGA-EE-A3J5-06A-11D-A20D-08.vcf", #"public_SKCM_TCGA-ER-A197-06A-32D-A197-08.vcf", "public_SKCM_TCGA-ER-A19O-06A-11D-A197-08.vcf", #"public_LUAD_TCGA-97-7938.vcf", "public_LUAD_TCGA-MP-A4TJ.vcf"), c(rep("Lung", 4), rep("Breast", 4)))
#colnames(sample_annotations) <- c("Sample_Names", "Tumor_Subtypes")
#write.table(sample_annotations, #"/Users/atgc/Documents/Aaron_Tools/BAGEL/inst/testdata/sample_annotations.txt", quote=FALSE, #row.names=FALSE, sep = "\t")
#
#sample_annotations <- read.table(system.file("testdata", "sample_annotations.txt", package = "BAGEL"), sep #= "\t", header=TRUE)
#init_sample_annotations(bay)
#add_sample_annotations(bay, sample_annotations, sample_column = "Sample_Names", columns_to_add = #"Tumor_Subtypes")
#create_snv96_table(bay, g)
#res <- find_signatures(bay, table_name = "SNV96", num_signatures = 3)
#plot_signatures(res)
#plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes")
```


```{r setup}
library("BAGEL")
```

```{r}
#Load MAF File
lusc_maf <- system.file("testdata", "public_TCGA.LUSC.maf", package = "BAGEL") 
lusc <- BAGEL::maf_file_to_dt(maf_file = lusc_maf)

#Load individual VCF Files
luad1_vcf <- system.file("testdata", "public_LUAD_TCGA-97-7938.vcf", 
                         package = "BAGEL")
luad1 <- BAGEL::vcf_file_to_dt(vcf_file = luad1_vcf)
luad2_vcf <- system.file("testdata", "public_LUAD_TCGA-MP-A4TJ.vcf", 
                         package = "BAGEL")
luad2 <- BAGEL::vcf_file_to_dt(vcf_file = luad2_vcf)

#Load Multiple VCF Files At Once using list files with regular expressions
melanoma_vcf <- list.files(system.file("testdata", package = "BAGEL"), 
                           pattern = glob2rx("*SKCM*vcf"), full.names = TRUE)
melanoma <- BAGEL::vcf_files_to_dt(vcf_files = melanoma_vcf)

#Combine all data.frames
dat <- rbind(lusc, melanoma, luad1, luad2)

#Choose genome build
#Automatically detect genome build
#g=BAGEL::select_genome("38")

#Create Object and add samples
bagel = new("bagel")
variants = dat
BAGEL::set_variants(bagel, variants)

#Or alternatively create an object with sample names included
bagel = new("bagel", variants = dat)

#Display what samples were added
BAGEL::get_sample_names(bagel)

#Create 96 Motif samples
#g=BAGEL::select_genome("19")
#BAGEL::create_tables(bagel, g)
g=BAGEL::select_genome("38")
#Generates the "SNV96" table
BAGEL::create_snv96_table(bagel, g)

#Generate signatures/weights
#We can use the bagel object we generated tables for using the SNV96 table
result=BAGEL::find_signatures(bagel, "SNV96", num_signatures = 4)

#We can name signatures based on prior knowledge
name_signatures(result, c("smoking", "uv", "apobec", "unknown"))

#Plot results
BAGEL::plot_signatures(result)
BAGEL::plot_exposures(result, proportional = TRUE)
BAGEL::plot_exposures(result, proportional = FALSE)

#Compare to COSMIC signatures by leaving the second result as default
BAGEL::compare_results(result, threshold = 0.8)

#List which signatures correspond to subtypes including "lung"
BAGEL::what_cosmic_v2_sigs("lung")

#Calculate posterior based on COSMIC signatures 4, 11, 12, 15
cosmic_post = infer_signatures(bagel = bagel, "SNV96", signatures_to_use = c(12, 4, 11, 15))
#cosmic_post = infer_signatures(bagel = bagel)
#to_use = as.numeric(substr(names(which(rowSums(cosmic_post@samples) > 10^-10)), 11, 12))
#to_use = as.numeric(substr(names(which(apply(cosmic_post@samples, 1, max) > 10^-10)), 11, 12))
#cosmic_post = infer_signatures(bagel = bagel, signatures_to_use = to_use)

#Calculate posterior based on all calculated signatures
our_sigs_post = infer_signatures(bagel = bagel, "SNV96", signatures = result@signatures)

#Plot results from posterior calculation
BAGEL::plot_signatures(cosmic_post)
BAGEL::plot_exposures(cosmic_post, proportional = TRUE)

BAGEL::plot_signatures(our_sigs_post)
BAGEL::plot_exposures(our_sigs_post, proportional = TRUE)

#Compare posterior results to each other
BAGEL::compare_results(cosmic_post, our_sigs_post, threshold = 0.70)

BAGEL::plot_sample_counts(bagel, "SNV96", get_sample_names(bagel)[1])

sample_annotations <- read.table(system.file("testdata", "sample_annotations.txt", package = "BAGEL"), sep = "\t", header=TRUE)
init_sample_annotations(bagel)
add_sample_annotations(bagel, sample_annotations, sample_column = "Sample_Names", columns_to_add = "Tumor_Subtypes")

#Add tumor type annotations to our samples
res <- find_signatures(bagel, table_name = "SNV96", num_signatures = 3)
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes")
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", proportional = FALSE)
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", by_group = FALSE)
plot_exposures_by_annotation(res, annotation = "Tumor_Subtypes", by_group = FALSE, proportional = FALSE)
```
